bip-0322.mediawiki

```diff
@@ -76,16 +88,33 @@ The <code>to_sign</code> transaction is:
 
 A full signature consists of the base64-encoding of the <code>to_sign</code> transaction in standard network serialisation once it has been signed.
 
+This signature, as-is, only proves a message is signed by the intended recipient of the given invoice address which creates the <code>message_challenge</code> script.
+It does not prove any control of bitcoins or having had sent any transactions.
+
+TODO: How does this interact with as-of-yet-unspecified "Silent Transactions"?
+
+TODO: Some invalid opcode to allow only in various proof types?
+
+TODO: A way for the initial signer to delegate to another scriptPubKey; needed for better privacy and CoinJoin/Lightning compatibility
+
 === Full (Proof of Funds) ===
 
-A signer may construct a proof of funds, demonstrating control of a set of UTXOs, by constructing a full signature as above, with the following modifications.
+A signer may construct a proof of funds, demonstrating ownership of an amount of bitcoins, by constructing a full signature as above, with the following modifications.
 
 * <code>message_challenge</code> is unused and shall be set to <code>OP_TRUE</code>
 * Similarly, <code>message_signature</code> is then empty.
-* All outputs that the signer wishes to demonstrate control of are included as additional inputs of <code>to_sign</code>, and their witness and scriptSig data should be set as though these outputs were actually being spent.
+* <code>vout[0].nValue</code> should be set to the desired amount of bitcoins to prove ownership of.
+* The signer must select UTXOs with values sum to at least the desired proof amount, and include them as additional inputs of <code>to_sign</code>. Their witness and scriptSig data should be set as though these outputs were actually being spent.
 
 Unlike an ordinary signature, validators of a proof of funds need access to the current UTXO set, to learn that the claimed inputs exist on the blockchain, and to learn their scriptPubKeys.
 
+This signature proves only that the message is signed by someone who owns the specified (in nValue) amount of bitcoins.
+It does not prove anything about the transactions or recipients thereof when the UTXOs were created;
+it does not prove control or ownership of the specific UTXOs;
```

Comment:
This seems contradictory.

Since the signature is over specific UTXOs, if the signature proves that the signer owns the specified amount, that implies that the signer owns those UTXOs, or at least has the ability to spend them as they have to provide valid scriptSigs and witnesses. If the signature does not prove that the signer has the ability to spend those specific UTXOs, then I don't see how the signature proves that they own the specified amount of Bitcoin.

Reply:
eg, MtGox should produce these signatures for their users, while still keeping funds mingled in separate hot/cold wallets. The hot UTXOs might be spent by any unrelated user, but in the meantime serve as proof of an available sum for another user. Two separate users might have proofs for different actual balances yet technically sign using the same UTXO(s).

Reply:
Should it say that the signature proves *control* over the UTXOs, but does not prove *ownership* or *unique control*?

Reply:
It doesn't prove control over the UTXOs, though, that's the point. It shows (perhaps not proves absolutely) ownership.

Reply:
I consider myself pretty knowledgeable, and you've been saying this for many years (about signmessage), and to be honest I've never fully understood what you mean.

My understanding is that signing a message (or a transaction) proves control over a UTXO, with the following caveats:
* it may not have been you who did the signing (e.g. you are a Gox user and Gox did the sig)
* it may not have been recent (Gox did it 10 years ago and since then has run into some trouble ... although I suppose you could include a recent headline and/or a challerger-provided nonce to help with this)
* even if it's recent, it is obsolete as soon as it's created (ok, you signed 1 second ago but your signing device caught fire at the same time)
* it may not indicate unique control (Gox may be signing with the same UTXOs for multiple users)
* it may not indicate general-purpose signing ability (maybe your hardware wallet will only sign messages, or only sign BIP322 provably-invalid transactions, etc., and won't actually sign real transactions)

Does this match your understanding? The word "ownership" feels nebulous to me and I'm not sure what you mean by it.

Reply:
For signed messages to have any meaning, it needs to indicate an intended purpose. Current signed messages indicate only that the message-signer receives/received what was sent to the address in question. They cannot prove control of a UTXO, because it is *expected, by design* for a third-party to produce the signature even if the recipient no longer has a claim to the UTXO received, and may never have had control of it in any case.

To prove control of funds, or control of UTXOs (if that indeed has a use case), a signature would need to intend to be usable for that purpose, distinguishable from current signatures, so third-party signing applications could refuse to produce such a signature when it is not valid.

Reply:
I don't understand why or since when stated purpose matters for cryptographic signatures

Reply:
???

Reply:
I don't understand your statement "For signed messages to have any meaning, it needs to indicate an intended purpose", since the mindset of the signer obviously doesn't affect the signature data

Reply:
DONE

Reply:
```suggestion
The meaning of the signature is not resolved by Bitcoin's blockchain.
```

Reply:
Without a meaning, it's not practical to implement correctly.

Reply:
I don't think this is true -- many things can be implemented correctly even though they have "no meaning" at the implementation layer. GPG signatures would be a close analogy to this (gnupg has no idea what the bytes you're signing are supposed to indicate, nor what your signature is supposed to mean), but this is also true of e.g. TCP.

In this case BIP322 is designed to represent a transaction, possibly with some extra non-transaction data embedded in it, and to demonstrate the ability to produce such a transaction. This can be mechanically defined and correctly implemented without agreeing on what the resulting transactions are supposed to mean.

Reply:
> Without a meaning, it's not practical to implement correctly.

I think the signatures do have meaning, but in the realm of human-mediated concerns, rather than the realm of Bitcoin blockchain transaction ordering.

I don't think that we can proscribe what people will use the signatures for.  It's still useful to have reliable code that allows these signatures to be checked.  This BIP can also warn about the limitations that people will encounter for whatever "L2" meaning they build, since BIP322-signature-double-spend protection won't be available.

Reply:
GPG only works with messages from one person. It doesn't have a concept of sender, receiver, or funds availability. These things all require distinct proofs, as the person who can produce such a signature may be different people entirely. Without a well-defined meaning at least to this degree, it isn't possible to use signed messages at all.

Reply:
Whether or not the script has been satisfied is meaningful.  How that is tied to a claim is meaning constructed outside the Bitcoin blockchain, by the parties that choose to use BIP322-signatures.  Unlike txs stored in the blockchain, these signatures don't prevent double spending their associated claims.  

Some people may agree that the ability to satisfy the script at one moment is still valuable, because they know the constraints of their L2 problem domain.

Reply:
ALL of this is outside the blockchain.

If a signature doesn't indicate a purpose, then the parties can't agree on a purpose, because one of them probably won't be able to produce a signature at all.

Reply:
Other metadata used to indicate the purpose can wrap the signature.

One approach to this, with well defined strucure for wrapping signatures, is the W3C Verifiable Credential Data Model, which is very specific about what a claim is: https://www.w3.org/TR/vc-data-model/#claims

Reply:
You can't even produce the signature until this purpose is defined...

Reply:
None of this is new.  Signatures go like this:

```
protocol indicating purpose, perhaps just words
--- begin signature ---
signature(protocol indicating purpose, perhaps just words)
```
 
In the case of the example I offered, the protocol indicating purpose is quite sophisticated.  I belive it's a compelling example that allows you no room to try to derail BIP322 on the grounds that all that meaning needs to be defined inside BIP322.

P.S.  This argument has moved beyond the proposed patch that started these replies, which I'm not arguing for or against, so I'll move on.